File: c++/heapTest.cpp
#include "test.h"
#include "heap.hpp"

using namespace std;

class Heap : public ::testing::Test
{
protected:
    vector<int> v{2, 3, 1, 4, 5};
};

TEST_F(Heap, sort_heap)
{
    vector<int> expected(v);
    std::make_heap(v.begin(), v.end());
    std::make_heap(expected.begin(), expected.end());
    std::sort_heap(expected.begin(), expected.end(), less<int>());
    my::sort_heap(v.begin(), v.end(), less<int>());
    EXPECT_EQ(expected, v);
}

TEST_F(Heap, is_heap)
{
    EXPECT_FALSE(my::is_heap(v.begin(), v.end()));
    std::make_heap(v.begin(), v.end());
    EXPECT_TRUE(my::is_heap(v.begin(), v.end()));
}

TEST_F(Heap, is_heap_until)
{
    EXPECT_EQ(std::is_heap_until(v.begin(), v.end()),
            my::is_heap_until(v.begin(), v.end()));
    std::make_heap(v.begin(), v.end());
    EXPECT_EQ(std::is_heap_until(v.begin(), v.end()),
            my::is_heap_until(v.begin(), v.end()));
}

TEST_F(Heap, make_heap)
{
    vector<int> duplicate = v;
    std::make_heap(duplicate.begin(), duplicate.end());
    my::make_heap(v.begin(), v.end());
    EXPECT_EQ(duplicate, v);
}

TEST_F(Heap, pop_heap)
{
    std::make_heap(v.begin(), v.end());
    vector<int> duplicate = v;
    std::pop_heap(duplicate.begin(), duplicate.end());
    my::pop_heap(v.begin(), v.end());
    EXPECT_EQ(duplicate, v);
}

TEST_F(Heap, push_heap)
{
    std::make_heap(v.begin(), v.end());
    v.push_back(11);
    EXPECT_FALSE(std::is_heap(v.begin(), v.end()));
    my::push_heap(v.begin(), v.end());
    EXPECT_TRUE(std::is_heap(v.begin(), v.end()));
}

File: php/HeapTest.php
<?php
namespace Algorithm;

class HeapTest extends \PHPUnit_Framework_TestCase
{
    private $ary;

    public function setUp() {
        $this->ary = [2, 3, 13, 8, 5, 7];
        $this->empty = [];
        $this->single = [1];
    }

    public function testIsHeap()
    {
        $this->assertTrue(isHeap([]));
        $this->assertTrue(isHeap([1]));
        $this->assertTrue(isHeap([2, 3, 13, 8, 5]));
        $this->assertFalse(isHeap($this->ary));
    }

    public function testIsHeapUntil()
    {
        $this->assertNull(isHeapUntil([]));
        $this->assertNull(isHeapUntil([1]));
        $this->assertNull(isHeapUntil([1, 2]));
        $this->assertNull(isHeapUntil([1, 2, 1]));
        $this->assertEquals(2, isHeapUntil([1, 2, 0]));
        $this->assertEquals(5, isHeapUntil($this->ary));
    }

    public function testMakeHeap()
    {
        makeHeap($this->empty);
        makeHeap($this->single);

        $this->assertFalse(isHeap($this->ary));
        makeHeap($this->ary);
        $this->assertTrue(isHeap($this->ary));

        $heap = [1, 2];
        makeHeap($heap);
        $this->assertEquals([1, 2], $heap);
        $heap = [2, 1];
        makeHeap($heap);
        $this->assertEquals([1, 2], $heap);
    }

    public function testPopHeap()
    {
        makeHeap($this->ary);
        $this->assertEquals(2, popHeap($this->ary));
        $this->assertTrue(isHeap($this->ary));
        $this->assertEquals(3, popHeap($this->ary));
        $this->assertTrue(isHeap($this->ary));

        $this->assertNull(popHeap($this->empty));
        $heap = [1, 2];
        $this->assertEquals(1, popHeap($heap));
    }

    public function testPushHeap()
    {
        makeHeap($this->ary);
        $this->assertNotContains(4, $this->ary);
        pushHeap($this->ary, 4);
        $this->assertContains(4, $this->ary);
        $this->assertTrue(isHeap($this->ary));

        $len = \count($this->ary);
        // Push null means doing nothing
        pushHeap($this->ary, null);
        $this->assertEquals($len, \count($this->ary));
    }

    public function testSortHeap()
    {
        sortHeap($this->ary);
        $this->assertEquals([2, 3, 5, 7, 8, 13], $this->ary);
        sortHeap($this->ary, function($a, $b){
            return $a > $b;
        });
        $this->assertEquals([13, 8, 7, 5, 3, 2], $this->ary);

        sortHeap($this->empty);
        sortHeap($this->single);
    }
}
